datafusion/physical-optimizer/src/wrap_leaves_cancellation.rs

```diff
@@ -76,7 +77,8 @@ impl WrapLeaves {
         plan: Arc<dyn ExecutionPlan>,
         yield_frequency: usize,
     ) -> Result<Transformed<Arc<dyn ExecutionPlan>>> {
-        let is_pipeline_breaker = plan.properties().emission_type == EmissionType::Final;
+        // todo this is a bit of a hack, we should probably have a more explicit way to handle
+        let is_pipeline_breaker = plan.properties().emission_type == EmissionType::Final|| plan.as_any().is::<FilterExec>();
```

Comment:
I was curious to see how this case could be handled. The consequence is that in many situations the yield logic will be injected unnecessarily causing the root stream to emit pending even though ready's are arriving at a steady pace. This is a bit at odds with the desire to keep performance overhead minimal.
I don't think there's a way to solve this external to `FilterExecStream` since you're trying to count the number of times it consecutively discarded full batches, not the number of times it polled its input.

Reply:
This code was removed by me, and we need to find a better solution for the filter and join corner cases.